### DISCUSSION TITLE
Converting a small parser from 0.9 to 1alpha - missing chain method

### DISCUSSION THREAD

**Discussion by dovreshef** [zesterer/chumsky#496] [Q&A] [answered]
Hi

I have the following parser:

```rust
 let byte = filter(char::is_ascii_hexdigit)
        .repeated()
        .exactly(2)
        .collect::<String>()
        .map(|v| u8::from_str_radix(&v, 16).unwrap());

    let hex = byte
        .chain(
            just::<_, _, Simple<_>>(':')
                .ignore_then(byte.separated_by(just(':')))
                .or(byte.repeated()),
        )
        .then_ignore(end());
```

This hex parser parse either a byte string formatted either as "1a:2b:3c:4e" or "1a2b3c4e".

I'm trying to move to the 1alpha version, and I see that the chain method on parser was removed.
What's the recommended alternative to replace it?

**Comment by zesterer** [comment]
You should use `map_slice` or `slice` instead, giving you access to the underlying input slice that was parsed by the parser, without any allocations at all.

**Reply by dovreshef** [reply]
Thanks, that saves me an allocation on the conversion to a u8.

Do you have any suggestion on how to replace the call to `chain`?

I saw that it got removed in this commit: https://github.com/zesterer/chumsky/commit/5c3c830912e2837e1a782c6bae991dcfa786a1ab 

It was more useful than `then` because with it I could create a `Vec` from a parser producing a single item, followed by a parser producing a list, without first creating the list and then inserting at the 0 location, which moves all the rest of the data..

**Reply by Zij-IT** [reply]
This begs the question of why `chain` was removed,  if it’s possible to either make it redundant, and if not, are there reasons not to reintroduce it

**Reply by zesterer** [reply]
It's not so much that it was removed, but that it was never reimplemented for `1.0`. Most uses of `chain` are covered by `[map_]slice` and those that do still exist should be possible with `map`. I'm not opposed to reintroducing it if there are strong feelings about reintroducing it though.

**Reply by dovreshef** [reply]
I'm not looking to bring it back, or opposing it, really. I don't feel like I can make an informed opinion about it.
I'm just looking for advice how can I emulate it using the existing combinators, without paying the small performance cost to collect it to a vec.

**Reply by zesterer** [reply]
This seems like it would do what you want:

```rs
let byte = any().filter(char::is_ascii_hexdigit)
        .repeated()
        .exactly(2)
        .map_slice(|s| u8::from_str_radix(s, 16).unwrap());

let hex = byte
	.repeated()
        .at_least(1)
	.collect::<Vec<_>>()
	.or(byte
		.separated_by(just(':'))
		.at_least(1)
		.collect());
```

FYI, `.then_ignore(end())` is no longer necessary in chumsky 1.0 :)

**Reply by Zij-IT** [reply]
I believe that would have to be `at_least(2)` for the first part of `hex`, because it will otherwise parse `"ab:cd"` incorrectly.

Assuming however that you do get it to produce the same output, it will still have a small difference from the original; It doesn't matter for this parser, but could when `byte` is more complex.

Given the input `"ab:cd"` the first `byte` `"ab"` is parsed twice in your solution, but only parsed once in the original. If we assume for a moment that `byte` is arbitrarily complex (like a language expression), then re-parsing it could be costly.

**Reply by dovreshef** [reply]
It also does not work, sadly..

A self containing, failing example:

```rust
use chumsky::prelude::*;

fn hex_parser<'src>() -> impl Parser<'src, &'src str, Vec<u8>, extra::Err<Rich<'src, char>>> {
    let byte = any()
        .filter(char::is_ascii_hexdigit)
        .repeated()
        .exactly(2)
        .map_slice(|v: &str| u8::from_str_radix(v, 16).unwrap());

    byte.repeated()
        .at_least(1)
        .collect::<Vec<_>>()
        .or(byte.separated_by(just(':')).at_least(1).collect())
}

fn main() {
    let v = hex_parser().parse("01020304").unwrap();
    assert_eq!(v, vec![0x01, 0x02, 0x03, 0x04]);
    let v = hex_parser().parse("01:02:03:04").unwrap();
    assert_eq!(v, vec![0x01, 0x02, 0x03, 0x04]);
}
```

The second assert fails, after it parses the first byte. It does not go to the or case since it has already succeed once with the first parser..

**Reply by zesterer** [reply, marked as answer]
Right, in that case you can just put the separated case first:

```rs
let byte = any().filter(char::is_ascii_hexdigit)
    .repeated()
    .exactly(2)
    .map_slice(|s| u8::from_str_radix(s, 16).unwrap());

let hex = byte
    .separated_by(just(':'))
    .at_least(2)
    .collect::<Vec<_>>()
    .or(byte
        .repeated()
        .at_least(1)
        .collect());
```

**Reply by dovreshef** [reply]
Thanks! that solved this.

And thanks for the library, it is very useful!
